File: providers/aws/src/main/java/cloudspec/aws/ec2/resource/nested/EC2NetworkAclEntry.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.aws.ec2.resource.nested;

import cloudspec.annotation.PropertyDefinition;
import software.amazon.awssdk.services.ec2.model.IcmpTypeCode;
import software.amazon.awssdk.services.ec2.model.NetworkAclEntry;

import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

public class EC2NetworkAclEntry {
    @PropertyDefinition(
            name = "cidr_block",
            description = "The IPv4 network range to allow or deny, in CIDR notation"
    )
    private final String cidrBlock;

    @PropertyDefinition(
            name = "egress",
            description = "Indicates whether the rule is an egress rule"
    )
    private final Boolean egress;

    @PropertyDefinition(
            name = "icmp_type_code",
            description = "The ICMP code. A value of -1 means all codes for the specified ICMP type"
    )
    private final Integer icmpTypeCode;

    @PropertyDefinition(
            name = "ipv6_cidr_block",
            description = "The IPv6 network range to allow or deny, in CIDR notation"
    )
    private final String ipv6CidrBlock;

    @PropertyDefinition(
            name = "port_range",
            description = "TCP or UDP protocols: The range of ports the rule applies to"
    )
    private final EC2PortRange portRange;

    @PropertyDefinition(
            name = "protocol",
            description = "The protocol number. A value of \"-1\" means all protocols"
    )
    private final String protocol;

    @PropertyDefinition(
            name = "rule_action",
            description = "Indicates whether to allow or deny the traffic that matches the rule",
            exampleValues = "allow | deny"
    )
    private final String ruleAction;

    public EC2NetworkAclEntry(String cidrBlock, Boolean egress, Integer icmpTypeCode, String ipv6CidrBlock,
                              EC2PortRange portRange, String protocol, String ruleAction) {
        this.cidrBlock = cidrBlock;
        this.egress = egress;
        this.icmpTypeCode = icmpTypeCode;
        this.ipv6CidrBlock = ipv6CidrBlock;
        this.portRange = portRange;
        this.protocol = protocol;
        this.ruleAction = ruleAction;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        EC2NetworkAclEntry that = (EC2NetworkAclEntry) o;
        return Objects.equals(cidrBlock, that.cidrBlock) &&
                Objects.equals(egress, that.egress) &&
                Objects.equals(icmpTypeCode, that.icmpTypeCode) &&
                Objects.equals(ipv6CidrBlock, that.ipv6CidrBlock) &&
                Objects.equals(portRange, that.portRange) &&
                Objects.equals(protocol, that.protocol) &&
                Objects.equals(ruleAction, that.ruleAction);
    }

    @Override
    public int hashCode() {
        return Objects.hash(cidrBlock, egress, icmpTypeCode, ipv6CidrBlock, portRange, protocol, ruleAction);
    }

    public static List<EC2NetworkAclEntry> fromSdk(List<NetworkAclEntry> networkAclEntries) {
        return Optional.ofNullable(networkAclEntries)
                .orElse(Collections.emptyList())
                .stream()
                .map(EC2NetworkAclEntry::fromSdk)
                .collect(Collectors.toList());
    }

    public static EC2NetworkAclEntry fromSdk(NetworkAclEntry networkAclEntry) {
        return Optional.ofNullable(networkAclEntry)
                .map(v -> new EC2NetworkAclEntry(
                        v.cidrBlock(),
                        v.egress(),
                        icmpTypeCodeFromSdk(v.icmpTypeCode()),
                        v.ipv6CidrBlock(),
                        EC2PortRange.fromSdk(v.portRange()),
                        v.protocol(),
                        v.ruleActionAsString()
                ))
                .orElse(null);
    }

    public static Integer icmpTypeCodeFromSdk(IcmpTypeCode icmpTypeCode) {
        return Optional.ofNullable(icmpTypeCode)
                .map(IcmpTypeCode::code)
                .orElse(null);
    }
}

File: providers/aws/src/main/kotlin/cloudspec/aws/ec2/nested/EC2NetworkAclEntry.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.aws.ec2.nested

import cloudspec.annotation.PropertyDefinition

data class EC2NetworkAclEntry(
        @property:PropertyDefinition(
                name = PROP_CIDR_BLOCK,
                description = PROP_CIDR_BLOCK_D
        )
        val cidrBlock: String?,

        @property:PropertyDefinition(
                name = PROP_EGRESS,
                description = PROP_EGRESS_D
        )
        val egress: Boolean?,

        @property:PropertyDefinition(
                name = PROP_ICMP_TYPE_CODE,
                description = PROP_ICMP_TYPE_CODE_D
        )
        val icmpTypeCode: Int?,

        @property:PropertyDefinition(
                name = PROP_IPV6_CIDR_BLOCK,
                description = PROP_IPV6_CIDR_BLOCK_D
        )
        val ipv6CidrBlock: String?,

        @property:PropertyDefinition(
                name = PROP_PORT_RANGE,
                description = PROP_PORT_RANGE_D
        )
        val portRange: EC2PortRange?,

        @property:PropertyDefinition(
                name = PROP_PROTOCOL,
                description = PROP_PROTOCOL_D
        )
        val protocol: String?,

        @property:PropertyDefinition(
                name = PROP_RULE_ACTION,
                description = PROP_RULE_ACTION_D,
                exampleValues = "allow | deny"
        )
        val ruleAction: String
) {
    companion object {
        const val PROP_CIDR_BLOCK = "cidr_block"
        const val PROP_CIDR_BLOCK_D = "The IPv4 network range to allow or deny, in CIDR notation"
        const val PROP_EGRESS = "egress"
        const val PROP_EGRESS_D = "Indicates whether the rule is an egress rule"
        const val PROP_ICMP_TYPE_CODE = "icmp_type_code"
        const val PROP_ICMP_TYPE_CODE_D = "The ICMP code. A value of -1 means all codes for the specified ICMP type"
        const val PROP_IPV6_CIDR_BLOCK = "ipv6_cidr_block"
        const val PROP_IPV6_CIDR_BLOCK_D = "The IPv6 network range to allow or deny, in CIDR notation"
        const val PROP_PORT_RANGE = "port_range"
        const val PROP_PORT_RANGE_D = "TCP or UDP protocols: The range of ports the rule applies to"
        const val PROP_PROTOCOL = "protocol"
        const val PROP_PROTOCOL_D = "The protocol number. A value of \"-1\" means all protocols"
        const val PROP_RULE_ACTION = "rule_action"
        const val PROP_RULE_ACTION_D = "Indicates whether to allow or deny the traffic that matches the rule"
    }
}
